File: Java/Graphs/TopologicalSortAndProblems/210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        //create graph
        ArrayList<ArrayList<Integer>> adjList = new ArrayList<>();
        for(int i=0;i<numCourses;i++){
            adjList.add(new ArrayList<>());
        }
        int m =prerequisites.length;
        for(int i=0;i<m;i++){
            adjList.get(prerequisites[i][1]).add(prerequisites[i][0]);
        }

        int indegree[]= new int[numCourses];
        for(int i=0;i<numCourses;i++){
            for(Integer neighbour:adjList.get(i)){
                indegree[neighbour]++;
            }
        }

        Queue<Integer> q = new LinkedList<>();
        for(int i=0;i<numCourses;i++){
            if(indegree[i] == 0){
                q.add(i);
            }
        }
        int ans[] = new int[numCourses];
        int i=0;
        while(!q.isEmpty()){
            int node = q.poll();
            ans[i++] = node;
            for(Integer neighbour : adjList.get(node)){
                indegree[neighbour]--;
                if(indegree[neighbour] == 0){
                    q.add(neighbour);
                }
            }
        }
        if(i != numCourses) return new int[]{};
        return ans;
    }
}

File: Python/Graphs/TopologicalSortAndProblems/210-course-schedule-ii/course-schedule-ii.py
class Solution:
    
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adjList : List[List[int]]  = []

        n = numCourses
        edges = prerequisites
        for i in range(n):
            adjList.append([])





        for i in range(len(edges)):
            u = edges[i][0]
            v = edges[i][1]

            adjList[v].append(u)



        indegree = [0]*n

        for i in range(n):
            for neighbour in adjList[i]:
                indegree[neighbour] +=1

        q = deque()




        for i in range(n):

            if indegree[i] == 0:
                q.append(i)

        cnt= 0
        
        topo = []

        while q:

            node = q.popleft()
            cnt+=1
            topo.append(node)

            for neighbour in adjList[node]:
                indegree[neighbour] -=1
                if indegree[neighbour] ==0:
                    q.append(neighbour)



        if cnt == n:
            return topo


        return  []